app/controllers/api/exchanges_products_controller.rb

```diff
@@ -0,0 +1,97 @@
+module Api
+  class ExchangesProductsController < Api::BaseController
+    skip_authorization_check only: [:index]
+
+    def index
+      @exchange = Exchange.find_by_id(params[:exchange_id])
+
+      render json: exchange_products,
+             each_serializer: Api::Admin::ForOrderCycle::SuppliedProductSerializer,
+             order_cycle: @exchange.order_cycle,
+             status: :ok
+    end
+
+    private
+
+    def exchange_products
+      if @exchange.incoming
+        products_for_incoming_exchange
+      else
+        products_for_outgoing_exchange
+      end
+    end
+
+    def products_for_incoming_exchange
+      supplied_products(@exchange.order_cycle, @exchange.sender)
+    end
+
+    def supplied_products(order_cycle, enterprise)
+      if order_cycle.prefers_product_selection_from_coordinator_inventory_only?
+        enterprise.supplied_products.visible_for(order_cycle.coordinator)
+      else
+        enterprise.supplied_products
+      end
+    end
+
+    def products_for_outgoing_exchange
+      products = []
+      enterprises_for_outgoing_exchange.each do |enterprise|
+        products.push( *supplied_products(@exchange.order_cycle, enterprise).to_a )
+
+        products.each do |product|
+          unless product_supplied_to_order_cycle?(product)
+            products.delete(product)
+          end
+        end
+      end
+      products
```

Comment:
this array work (*) was just a literal conversion of the angular algorithm and is transformed, in a later commit, in a normal AR query with relations.